File: api/network.go
package api

// Network provides access to core network APIs
type Network interface {
	Register(name string) (PortHandler, error)
	Connect(identity Identity, port string) (Stream, error)
	Identity() Identity
}

// PortHandler is a handler for a locally registered port
type PortHandler interface {
	Requests() <-chan ConnectionRequest
	Close() error
}

// ConnectionRequest represents a connection request sent to a port
type ConnectionRequest interface {
	Caller() Identity
	Query() string
	Accept() Stream
	Reject()
}

// Stream represents a bidirectional stream
type Stream interface {
	Read(p []byte) (n int, err error)
	Write(p []byte) (n int, err error)
	Close() error
}

type Identity string

File: mobile/android/client/src/main/java/cc/cryptopunks/astral/net/Network.kt
package cc.cryptopunks.astral.net

import java.io.Closeable
import java.io.InputStream
import java.io.OutputStream

/**
 * Network provides access to core network APIs
 */
interface Network {
    fun query(port: String, identity: String = ""): Stream
    fun register(port: String): Port
    fun identity(): String

    object Exception {
        class Query(port: String, identity: String, cause: Throwable) :
            kotlin.Exception("$port $identity", cause)

        class Register(port: String, cause: Throwable) :
            kotlin.Exception(port, cause)

        class Resolve(name: String, cause: Throwable) :
            kotlin.Exception(name, cause)
    }
}

/**
 * PortHandler is a handler for a locally registered port
 */
interface Port : Closeable {
    operator fun next(): () -> Connection

    class Exception(
        port: String,
        cause: Throwable,
    ) : kotlin.Exception(port, cause)
}

/**
 * ConnectionRequest represents a connection request sent to a port
 */
interface Connection {
    fun accept(): Stream
    fun caller(): String
    fun query(): String
    fun reject()


    sealed class Exception(
        connection: Connection,
        cause: Throwable,
    ) :
        kotlin.Exception("${connection.query()} ${connection.caller()}", cause) {
        class Accept(connection: Connection, cause: Throwable) : Exception(connection, cause)
        class Reject(connection: Connection, cause: Throwable) : Exception(connection, cause)
    }
}

/**
 * Stream represents a bidirectional stream
 */
interface Stream : Closeable {
    fun read(buffer: ByteArray): Int
    fun write(buffer: ByteArray): Int
    val input: InputStream
    val output: OutputStream
}
